Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+HTTP1State.swift

```diff
@@ -0,0 +1,694 @@
+                    case (false, .some((_, let existingMatchReturn))):
+                        // We don't require a specific eventLoop. For this reason we want to pick a
+                        // matching eventLoop that has been idle the shortest.
+                        if connReturn > existingMatchReturn {
+                            goodMatch = (index, connReturn)
+                        }
+                    }
+                }
+            }
+
+            // if we found an eventLoopMatch, we can execute the request right away
+            if let (index, _) = eventLoopMatch {
+                assert(self.waiters.isEmpty, "If a connection is available, why are there any waiters")
+                var connectionState = self.connections[index]
+                let connection = connectionState.lease()
+                self.connections[index] = connectionState
+                return .init(
+                    .executeRequest(request, connection, cancelWaiter: nil),
+                    .cancelTimeoutTimer(connectionState.connectionID)
+                )
+            }
+
+            // if we found a good match, let's use this
+            if let (index, _) = goodMatch {
+                assert(self.waiters.isEmpty, "If a connection is available, why are there any waiters")
+                if !required {
+                    var connectionState = self.connections[index]
+                    let connectionID = connectionState.connectionID
+                    let connection = connectionState.lease()
+                    self.connections[index] = connectionState
+                    return .init(
+                        .executeRequest(request, connection, cancelWaiter: nil),
+                        .cancelTimeoutTimer(connectionID)
+                    )
+                } else {
+                    assert(self.connections.count - self.maximumConcurrentConnections == 0)
+                    var oldConnectionState = self.connections[index]
+                    let newConnectionID = self.idGenerator.next()
+                    let newWaiter = Waiter(request: request)
+                    self.connections[index] = .init(connectionID: newConnectionID, eventLoop: prefferedEL, waiter: newWaiter)
+                    return .init(
+                        .scheduleWaiterTimeout(newWaiter.requestID, request, on: prefferedEL),
+                        .replaceConnection(oldConnectionState.close(), with: newConnectionID, on: prefferedEL)
+                    )
+                }
+            }
+
+            // we didn't find any match at all... Let's create a new connection, if there is room
+            // left
+            if self.connections.count < self.maximumConcurrentConnections {
+                let newConnectionID = self.idGenerator.next()
+                let newWaiter = Waiter(request: request)
+                self.connections.append(.init(connectionID: newConnectionID, eventLoop: prefferedEL, waiter: newWaiter))
```

Comment:
It is unwise to force a waiter to own a _specific_ connection attempt. Instead, you want to queue up waiters and serve them using a FIFO as connections become available. This helps prevent issues where either a) waiter timeouts throw away connections we could use to serve other waiters, or b) problems where connections become active in a different order to the order of waiters arriving, risking more timeouts than necessary.